core/drivers/amd/ps_gpio_driver.c

```diff
@@ -0,0 +1,249 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2002-2021 Xilinx, Inc.  All rights reserved.
+ * Copyright (c) 2022 Foundries.io Ltd. (jorge@foundries.io)
+ * Copyright (c) 2024-2025, Advanced Micro Devices, Inc. All rights reserved.
+ *
+ */
+
+#include <assert.h>
+#include <drivers/gpio.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_mmu.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <trace.h>
+#include <util.h>
+
+#include "gpio_private.h"
+
+static struct amd_gbank_data ps_bank = {
+	.label = "ps_gpio",
+	.ngpio = 58,
+	.max_bank = PS_BANK_MAX,
+	.bank_min[0] = 0,
+	.bank_max[0] = 25,
+	.bank_min[3] = 26,
+	.bank_max[3] = 57,
+};
+
+/* Standard GPIO Operations */
+static enum gpio_level ps_gpio_get_value(struct gpio_chip *chip,
+					 unsigned int gpio_pin)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	amd_gpio_get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if ((io_read32(ps->vbase + DATA_RO_OFFSET(bank)) >> pin) & BIT(0))
+		return GPIO_LEVEL_HIGH;
+
+	return GPIO_LEVEL_LOW;
+}
+
+static void ps_gpio_set_value(struct gpio_chip *chip,
+			      unsigned int gpio_pin,
+			      enum gpio_level level)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	uint32_t offset = 0;
+	uint32_t lvl = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	amd_gpio_get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if (pin < GPIO_NUM_MAX) {
+		offset = DATA_LSW_OFFSET(bank);
+	} else {
+		pin -= GPIO_NUM_MAX;
+		offset = DATA_MSW_OFFSET(bank);
+	}
+
+	/* Explicitly compare the enum value */
+	if (level == GPIO_LEVEL_HIGH)
+		lvl = 1;
+	else
+		lvl = 0;
+
+	lvl = ~BIT32(pin + GPIO_NUM_MAX) &
+		(SHIFT_U32(lvl, pin) | GPIO_UPPER_MASK);
+
+	io_write32(ps->vbase + offset, lvl);
+}
+
+static enum gpio_dir ps_gpio_get_dir(struct gpio_chip *chip,
+				     unsigned int gpio_pin)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	amd_gpio_get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if (io_read32(ps->vbase + DIRM_OFFSET(bank)) & BIT(pin))
+		return GPIO_DIR_OUT;
+
+	return GPIO_DIR_IN;
+}
+
+static void ps_gpio_set_dir(struct gpio_chip *chip,
+			    unsigned int gpio_pin,
+			    enum gpio_dir direction)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	amd_gpio_get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if (direction == GPIO_DIR_OUT) {
+		/* set the GPIO pin as output */
+		io_setbits32(ps->vbase + DIRM_OFFSET(bank), BIT(pin));
```

Comment:
Seen the few  `io_setbits32()`/`io_clrbits32()` calls in this function, I think it needs some kind of concurrent access protection.

Ditto for `ps_gpio_set_intr()`.